File: tests/snapshots/enum_payloads.rs
// Enum variants can now carry any payload — structs, several values, even a
// `Vec` (which also lets an enum hold a collection of things).

#[derive(Debug, Clone)]
struct Point {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone)]
#[allow(dead_code)]
enum Shape {
    Dot(Point),
    Segment(Point, Point),
    Blob(Vec<Point>),
    Empty,
}
impl std::fmt::Display for Shape {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

fn describe(s: Shape) -> Result<String, String> {
    match s {
        Shape :: Dot ( p ) => {
            return Ok(format!("dot at {},{}", p.x, p.y));
        }
        Shape :: Segment ( a , b ) => {
            return Ok(format!("segment {} to {}", a.x, b.x));
        }
        Shape :: Blob ( pts ) => {
            return Ok(format!("blob of {} points", pts.len()));
        }
        Shape :: Empty => {
            return Ok("nothing".to_string());
        }
    }
}

fn vbr_main() -> Result<(), String> {
    println!("{}", describe(Shape::Dot(Point { x: 1.0, y: 2.0 }))?);
    println!("{}", describe(Shape::Segment(Point { x: 1.0, y: 2.0 }, Point { x: 5.0, y: 6.0 }))?);
    let mut cloud: Vec<Point> = Vec::new();
    cloud.push(Point { x: 1.0, y: 2.0 });
    cloud.push(Point { x: 5.0, y: 6.0 });
    println!("{}", describe(Shape::Blob(cloud))?);
    println!("{}", describe(Shape::Empty)?);
    Ok(())
}

fn main() {
    if let Err(error) = vbr_main() {
        eprintln!("Error: {error}");
        std::process::exit(1);
    }
}

File: tests/snapshots/enum_payloads.py
# Enum variants can now carry any payload — structs, several values, even a
# `Vec` (which also lets an enum hold a collection of things).

import sys
from dataclasses import dataclass

@dataclass
class Ok:
    value: object

@dataclass
class Err:
    error: object

def _vb(x):
    if isinstance(x, bool):
        return "true" if x else "false"
    if isinstance(x, float) and x.is_integer():
        return str(int(x))
    return str(x)

@dataclass
class Point:
    x: float
    y: float

class Shape:
    pass

@dataclass
class Dot(Shape):
    f0: Point

@dataclass
class Segment(Shape):
    f0: Point
    f1: Point

@dataclass
class Blob(Shape):
    f0: list[Point]

@dataclass
class Empty(Shape):
    pass

def describe(s: Shape) -> str:
    _m0 = s
    match _m0:
        case Dot(p):
            return Ok(f"dot at {_vb(p.x)},{_vb(p.y)}")
        case Segment(a, b):
            return Ok(f"segment {_vb(a.x)} to {_vb(b.x)}")
        case Blob(pts):
            return Ok(f"blob of {_vb(len(pts))} points")
        case Empty():
            return Ok('nothing')

def main():
    _t0 = describe(Dot(Point(x=1.0, y=2.0)))
    if isinstance(_t0, Err):
        print(f"Error: {_t0.error}", file=sys.stderr)
        raise SystemExit(1)
    print(_vb(_t0.value))
    _t1 = describe(Segment(Point(x=1.0, y=2.0), Point(x=5.0, y=6.0)))
    if isinstance(_t1, Err):
        print(f"Error: {_t1.error}", file=sys.stderr)
        raise SystemExit(1)
    print(_vb(_t1.value))
    cloud: list[Point] = []
    cloud.append(Point(x=1.0, y=2.0))
    cloud.append(Point(x=5.0, y=6.0))
    _t2 = describe(Blob(cloud))
    if isinstance(_t2, Err):
        print(f"Error: {_t2.error}", file=sys.stderr)
        raise SystemExit(1)
    print(_vb(_t2.value))
    _t3 = describe(Empty())
    if isinstance(_t3, Err):
        print(f"Error: {_t3.error}", file=sys.stderr)
        raise SystemExit(1)
    print(_vb(_t3.value))


if __name__ == "__main__":
    main()
